Raspberry Pi Camera V3, Raspberry Pi 5, Ubuntu Server 24.04, ROS2 Jazzy

Hello,

Not really an issue, but I thought it may be helpful for those needing to use the Rpi Cam V3 (Sony IMX708) on Raspberry Pi 5. I tested it on ROS2 Jazzy and Ubuntu Server 24.04. 

To use the camera we need to build [Raspberry Pi's `libcamera` fork](https://github.com/raspberrypi/libcamera/) and `camera_ros` from source. The recent commits of the fork have caused some `isArray_` assertion error shown at the bottom of this post. To fix the problem, simply switch back to a working commit. The following commands show how both repos can be built from source by switching to an earlier commit for the `libcamera` fork (modified from this repo's README):

```bash
# create workspace
mkdir -p ~/camera_ws/src
cd ~/camera_ws/src

# check out libcamera
sudo apt -y install python3-colcon-meson  python3-ply # ply is required for ubuntu server.
# Option B: raspberrypi fork with support for newer camera modules
git clone https://github.com/raspberrypi/libcamera.git

# check out libcamera's Aug 25 commit with SHA bfd68f786964636b09f8122e6c09c230367390e7
# The git reset output should be: bfd68f78 libcamera: clock_recovery: Use nanoseconds in addSample()
cd ~/camera_ws/src/libcamera
git reset --hard bfd68f786964636b09f8122e6c09c230367390e7
cd ~/camera_ws/src

# check out this camera_ros repository
git clone https://github.com/christianrauch/camera_ros.git

# resolve binary dependencies and build workspace
source /opt/ros/$ROS_DISTRO/setup.bash
cd ~/camera_ws/
rosdep install -y --from-paths src --ignore-src --rosdistro $ROS_DISTRO --skip-keys=libcamera
colcon build --event-handlers=console_direct+
```

# `isArray_` assertion failed with recent commits of `libcamera`

If the particular commit in `libcamera` is not used, the error will occur. The terminal output below shows the error, and is generated by running `ros2 run camera_ros camera_node` with the most recent commit (as of writing). Setting different ROS2 parameters for `camera_node` consistently yielded the same error.

```bash
[0:18:17.459909805] [3856]  INFO Camera camera_manager.cpp:340 libcamera v0.6.0+rpt20251202
[0:18:17.480914313] [3871]  INFO RPI pisp.cpp:720 libpisp version v1.3.0 9ba67e6680f0 13-12-2025 (08:24:06)
[0:18:17.569594645] [3871]  INFO IPAProxy ipa_proxy.cpp:180 Using tuning file /home/ece26/camera_ws/install/libcamera/share/libcamera/ipa/rpi/pisp/imx708.json
[0:18:17.581873782] [3871]  INFO Camera camera_manager.cpp:223 Adding camera '/base/axi/pcie@120000/rp1/i2c@88000/imx708@1a' for pipeline handler rpi/pisp
[0:18:17.581905690] [3871]  INFO RPI pisp.cpp:1181 Registered camera /base/axi/pcie@120000/rp1/i2c@88000/imx708@1a to CFE device /dev/media0 and ISP device /dev/media1 using PiSP variant BCM2712_D0
[INFO] [1765614332.175562731] [camera]: 
>> cameras:
   0: imx708 (/base/axi/pcie@120000/rp1/i2c@88000/imx708@1a)
[WARN] [1765614332.175710715] [camera]: no camera selected, using default: "/base/axi/pcie@120000/rp1/i2c@88000/imx708@1a"
[WARN] [1765614332.175739011] [camera]: set parameter 'camera' to silence this warning
[0:18:17.582558735] [3871]  WARN V4L2 v4l2_pixelformat.cpp:346 Unsupported V4L2 pixel format RPBP
[INFO] [1765614332.176370131] [camera]: 
>> stream formats:
   - R8 (32x32 - 4608x2592)
   - R16 (32x32 - 4608x2592)
   - MONO_PISP_COMP1 (32x32 - 4608x2592)
   - NV21 (32x32 - 4608x2592)
   - SBGGR8 (32x32 - 4608x2592)
   - SBGGR16 (32x32 - 4608x2592)
   - BGGR_PISP_COMP1 (32x32 - 4608x2592)
   - XBGR8888 (32x32 - 4608x2592)
   - BGR888 (32x32 - 4608x2592)
   - RGB888 (32x32 - 4608x2592)
   - XRGB8888 (32x32 - 4608x2592)
   - SGBRG16 (32x32 - 4608x2592)
   - GBRG_PISP_COMP1 (32x32 - 4608x2592)
   - SGRBG16 (32x32 - 4608x2592)
   - GRBG_PISP_COMP1 (32x32 - 4608x2592)
   - SRGGB16 (32x32 - 4608x2592)
   - RGGB_PISP_COMP1 (32x32 - 4608x2592)
   - BGR161616 (32x32 - 4608x2592)
   - RGB161616 (32x32 - 4608x2592)
   - SRGGB8 (32x32 - 4608x2592)
   - SGRBG8 (32x32 - 4608x2592)
   - SGBRG8 (32x32 - 4608x2592)
   - YUYV (32x32 - 4608x2592)
   - UYVY (32x32 - 4608x2592)
[WARN] [1765614332.176474595] [camera]: no pixel format selected, auto-selecting: "XRGB8888"
[WARN] [1765614332.176495040] [camera]: set parameter 'format' to silence this warning
[INFO] [1765614332.176563615] [camera]: 
>> XRGB8888 format sizes:
   - 160x120
   - 240x160
   - 320x240
   - 400x240
   - 480x320
   - 640x360
   - 640x480
   - 720x480
   - 768x480
   - 854x480
   - 720x576
   - 800x600
   - 960x540
   - 1024x576
   - 960x640
   - 1024x600
   - 1024x768
   - 1280x720
   - 1152x864
   - 1280x800
   - 1360x768
   - 1366x768
   - 1440x900
   - 1280x1024
   - 1536x864
   - 1280x1080
   - 1600x900
   - 1400x1050
   - 1680x1050
   - 1600x1200
   - 1920x1080
   - 2048x1080
   - 1920x1200
   - 2160x1080
   - 2048x1152
   - 2560x1080
   - 2048x1536
   - 2560x1440
   - 2560x1600
   - 3840x1080
   - 2960x1440
   - 3440x1440
   - 2560x2048
   - 3200x1800
   - 3840x1600
   - 3200x2048
   - 3200x2400
   - 3840x2160
   - 4096x2160
   - 3840x2400
[WARN] [1765614332.176688857] [camera]: no dimensions selected, auto-selecting: "800x600"
[WARN] [1765614332.176708468] [camera]: set parameter 'width' or 'height' to silence this warning
[0:18:17.583373098] [3856]  INFO Camera camera.cpp:1215 configuring streams: (0) 800x600-XRGB8888/sRGB
[0:18:17.583530655] [3871]  INFO RPI pisp.cpp:1485 Sensor: /base/axi/pcie@120000/rp1/i2c@88000/imx708@1a - Selected sensor format: 1536x864-SBGGR10_1X10/RAW - Selected CFE format: 1536x864-PC1B/RAW
[INFO] [1765614332.177472552] [camera]: camera "/base/axi/pcie@120000/rp1/i2c@88000/imx708@1a" configured with 800x600-XRGB8888/sRGB stream
[INFO] [1765614332.177693166] [camera]: using default calibration URL
[INFO] [1765614332.177725055] [camera]: camera calibration URL: file:///home/ece26/.ros/camera_info/imx708__base_axi_pcie_120000_rp1_i2c_88000_imx708_1a_800x600.yaml
[ERROR] [1765614332.177791278] [camera_calibration_parsers]: Unable to open camera calibration file [/home/ece26/.ros/camera_info/imx708__base_axi_pcie_120000_rp1_i2c_88000_imx708_1a_800x600.yaml]
[WARN] [1765614332.177858168] [camera]: Camera calibration file /home/ece26/.ros/camera_info/imx708__base_axi_pcie_120000_rp1_i2c_88000_imx708_1a_800x600.yaml not found
[WARN] [1765614332.178766939] [camera]: unsupported control: AfWindows (37)
camera_node: /home/ece26/camera_ws/install/libcamera/include/libcamera/libcamera/controls.h:204: T libcamera::ControlValue::get() const [with T = libcamera::Span<const long int, 18446744073709551615>; typename std::enable_if<(libcamera::details::is_span<U>::value || std::is_same<std::basic_string_view<char>, typename std::remove_cv< <template-parameter-1-1> >::type>::value), std::nullptr_t>::type <anonymous> = nullptr]: Assertion `isArray_' failed.
[ros2run]: Aborted
```

This is clearly a `libcamera` problem, but I don't know enough about the code or have enough time to raise an issue there. Since I am using the camera for projects using `camera_ros`, I thought it may be useful to post a solution here. Please feel free to close the issue.

--- a/src/clamp.cpp
+++ b/src/clamp.cpp
@@ -106,14 +106,30 @@ libcamera::ControlValue
 clamp_array(const libcamera::ControlValue &value, const libcamera::ControlValue &min,
             const libcamera::ControlValue &max)
 {
+  assert(min.isArray() == max.isArray());
+  assert(min.numElements() == max.numElements());
+
   const libcamera::Span<const T> v = value.get<libcamera::Span<const T>>();
-  const libcamera::Span<const T> a = min.get<libcamera::Span<const T>>();
-  const libcamera::Span<const T> b = max.get<libcamera::Span<const T>>();
 
   std::vector<T> vclamp(v.size());
 
-  for (size_t i = 0; i < v.size(); i++)
-    vclamp[i] = std::clamp(v[i], a[i], b[i]);
+  if (min.isArray() == true && max.isArray() == true) {
+    // clamp individual elements with their individual min/max values
+    const libcamera::Span<const T> a = min.get<libcamera::Span<const T>>();
+    const libcamera::Span<const T> b = max.get<libcamera::Span<const T>>();
+    for (size_t i = 0; i < v.size(); i++)
+      vclamp[i] = std::clamp(v[i], a[i], b[i]);
+  }
+  else if (min.isArray() == false && max.isArray() == false) {
+    // clamp array elements via scalar min/max value
+    const T a = min.get<const T>();
+    const T b = max.get<const T>();
+    for (size_t i = 0; i < v.size(); i++)
+      vclamp[i] = std::clamp(v[i], a, b);
+  }
+  else {
+    throw should_not_reach();
+  }
 
   return libcamera::ControlValue(libcamera::Span<const T>(vclamp));
 }
@@ -140,7 +156,7 @@ libcamera::ControlValue
 clamp(const libcamera::ControlValue &value, const libcamera::ControlValue &min,
       const libcamera::ControlValue &max)
 {
-  if (min.type() != max.type())
+  if (min.type() != max.type() || min.isArray() != max.isArray() || min.numElements() != max.numElements())
     throw invalid_conversion("minimum (" + std::to_string(min.type()) + ") and maximum (" +
                              std::to_string(max.type()) + ") types mismatch");
 

